documentation/istio_keycloak_integration.asciidoc

```diff
@@ -0,0 +1,98 @@
+
+kubectl apply -f k8s/deployment.yaml
+kubectl apply -f k8s/service.yaml
+```
+
+You now should see a response from the application when using curl to reach the URL http://demo-quarkus.default:8080/animals.
+```
+kubectl exec $(kubectl get pod -l app=demo-quarkus -o jsonpath={.items..metadata.name}) -c demo-quarkus -- curl http://demo-quarkus.default:8080/animals
+Response: {"totalElements":0,"number":0,"size":100,"totalPages":0,"stream":[]}
+```
+
+This will only work from the Kubernetes cluster, because we use the name of the Kubernetes service to reach the service. To explode the application use:
+```
+kubectl apply -f k8s/istio/demo-quarkus-gateway.yaml
+```
+You should now able to open the URL http://demo-quarkus.localhost/animals in your browser.
+
+### Deploy and configure Keycloak
+We use Keycloak for identity management. Apply the following files to your cluster to start Keycloak and explode it via Istio ingress gateway.
+```
+kubectl apply -f k8s/keycloak/keycloak.yaml
+kubectl apply -f k8s/keycloak/keycloak-gateway.yaml
+```
+
+You should be able to open http://keycloak-demo-quarkus.localhost/auth/ in the browser. Open the Keycloak administration console and log in with `admin` as username and password.
+Add a new realm by selecting the realm-export.json file.
+
+Click on `Groups` and create a new group `User`. Then click on `Roles` and create the following roles:
+```
+admin, user, demo-quarkus.FindObject, demo-quarkus.SaveObject, demo-quarkus.DeleteObject
+```
+Now create a new user called `demo` and add it to the group you created.
+Finally, open the client `demo-quarkus-cli`, click on the tab Mapper and add the group mapper to the client. Your Keyloak is now fully configured.
+
+### Test the application
+Now it is time to test the application. Run the following command:
+```
+kubectl exec $(kubectl get pod -l app=demo-quarkus -o jsonpath={.items..metadata.name}) -c demo-quarkus -- curl http://demo-quarkus.default:8080/animals
+```
+You should get a valid response from the application. But there are no animals in our database at the moment. So let's try to create an animal. To do this, run the following command:
+```
+kubectl exec $(kubectl get pod -l app=demo-quarkus -o jsonpath={.items..metadata.name}) -c demo-quarkus -- curl -H "Content-Type: application/json" --request POST --data '{"name": "dog", "basicInfo": "home pet", "numberOfLegs":4}' http://demo-quarkus.default:8080/animals -i
+```
+You will get an error message 401 Unauthorized. This is because this operation is secured with the role `demo-quarkus.SaveObject`. You can only access this operation if you pass a valid JWT token in the request header. So add the role to the user in Keycloak and run the following command to get the token.
+```
+TOKEN=$(curl -d 'client_id=demo-quarkus-cli' -d 'username=demo' -d 'password=demo' -d 'grant_type=password' 'http://keycloak-demo-quarkus.localhost/auth/realms/demo-quarkus/protocol/openid-connect/token' | jq ".access_token" -r)
```

Comment:
request token with client secret too. `'client_secret=7745de1c-a85e-4bfe-97fe-42dc2b0b3f65'`

Reply:
You only need to pass the client secret if 'Access Type' of the Keycloak client is set to `confidential`. In the json file of the realm, the client 'demo-quarkus-cli' is specified as public client.